File: src/mvmctl/core/network/_resolver.py
"""Network resolution helpers."""

from __future__ import annotations

from dataclasses import dataclass

from mvmctl.core._internal._enrichment import RelationEnricher
from mvmctl.core.network._iptables_resolver import IPTablesRuleResolver
from mvmctl.core.network._lease_resolver import NetworkLeaseResolver
from mvmctl.core.network._repository import NetworkRepository
from mvmctl.exceptions import NetworkNotFoundError
from mvmctl.models.network import NetworkItem

__all__ = [
    "NetworkResolver",
    "NetworkResolveResult",
]


@dataclass
class NetworkResolveResult:
    items: list[NetworkItem]
    errors: list[str]
    exit_code: int


class NetworkResolver:
    """Resolver for network configuration."""

    RELATIONS: dict[str, tuple[str, type, str]] = {
        "leases": ("id", NetworkLeaseResolver, "list_by_network_id"),
        "iptables_rules": ("id", IPTablesRuleResolver, "list_by_network_id"),
    }

    def __init__(
        self,
        repo: NetworkRepository | None = None,
        *,
        include: list[str] | None = None,
    ) -> None:
        self._repo = repo if repo is not None else NetworkRepository()
        self._include = include

    def _enrich(self, networks: list[NetworkItem]) -> list[NetworkItem]:
        """Enrich networks with relations if include is set."""
        if self._include and networks:
            RelationEnricher().enrich(
                networks, self._include, self.RELATIONS, self._repo._db
            )
        return networks

    def by_id(self, network_id: str) -> NetworkItem:
        """Resolve network by ID prefix."""
        matches = self._repo.find_by_prefix(network_id)
        if len(matches) == 0:
            raise NetworkNotFoundError(f"Network not found: {network_id}")
        if len(matches) > 1:
            raise NetworkNotFoundError(f"Network ID is ambiguous: {network_id}")
        return self._enrich(matches)[0]

    def by_name(self, name: str) -> NetworkItem:
        """Resolve network by name."""
        network = self._repo.get_by_name(name)
        if network is None:
            raise NetworkNotFoundError(f"Network not found: {name}")
        return self._enrich([network])[0]

    def get_default(self) -> NetworkItem | None:
        """Resolve the default network, or None if not set."""
        network = self._repo.get_default()
        if network is None:
            return None
        return self._enrich([network])[0]

    def resolve(self, value: str) -> NetworkItem:
        """Resolve network by name or ID prefix."""
        try:
            network = self.by_name(value)
        except NetworkNotFoundError:
            network = self.by_id(value)
        return network

    def resolve_many(self, identifiers: list[str]) -> NetworkResolveResult:
        """Resolve multiple network identifiers by name or id."""
        # Deduplicate identifiers while preserving order
        seen_inputs: set[str] = set()
        unique_ids: list[str] = []
        for ident in identifiers:
            if ident not in seen_inputs:
                seen_inputs.add(ident)
                unique_ids.append(ident)

        items: list[NetworkItem] = []
        errors: list[str] = []
        resolved_ids: set[str] = set()

        for identifier in unique_ids:
            try:
                item = self.resolve(identifier)
                if item.id not in resolved_ids:
                    resolved_ids.add(item.id)
                    items.append(item)
            except Exception as e:
                errors.append(f"{identifier}: {e}")

        items = self._enrich(items)

        exit_code = 1 if errors and not items else 0
        return NetworkResolveResult(
            items=items, errors=errors, exit_code=exit_code
        )

File: internal/core/network/resolver.go
package network

import (
	"context"
	"fmt"

	"mvmctl/internal/infra"
	"mvmctl/internal/lib/model"
	"mvmctl/pkg/errs"
)

// ResolveResult holds the result of multi-resolve.
type ResolveResult struct {
	Items    []*model.NetworkItem
	Errors   []string
	ExitCode int
}

// NetworkEnrichFunc is a callback for enriching networks with relations.
// Set by the API layer during wiring to avoid circular imports.
type NetworkEnrichFunc func(ctx context.Context, networks []*model.NetworkItem) ([]*model.NetworkItem, error)

// Resolver resolves network identifiers.
type Resolver struct {
	repo     Repository
	include  []string
	enrichFn NetworkEnrichFunc
}

func NewResolver(repo Repository, include []string) *Resolver {
	return &Resolver{repo: repo, include: include}
}

// SetEnrichFunc sets the enrichment function called after each resolution.
// Must be set for cross-domain relation enrichment (e.g., populating leases,
// firewall rules, and VMs).
func (r *Resolver) SetEnrichFunc(fn NetworkEnrichFunc) {
	r.enrichFn = fn
}

// Enrich enriches networks with relations if include is set.
func (r *Resolver) enrich(ctx context.Context, networks []*model.NetworkItem) []*model.NetworkItem {
	if r.include == nil || len(r.include) == 0 || len(networks) == 0 {
		return networks
	}
	if r.enrichFn != nil {
		enriched, err := r.enrichFn(ctx, networks)
		if err == nil {
			return enriched
		}
	}
	return networks
}

// EnrichWithRelations loads relations for a resolved network.
// This is the public entry point for the enricher package to call.
func (r *Resolver) EnrichWithRelations(ctx context.Context, networks []*model.NetworkItem) []*model.NetworkItem {
	return r.enrich(ctx, networks)
}

func (r *Resolver) ByID(ctx context.Context, networkID string, includeDeleted ...bool) (*model.NetworkItem, error) {
	matches, err := r.repo.FindByPrefix(ctx, networkID, includeDeleted...)
	if err != nil {
		return nil, err
	}
	if len(matches) == 0 {
		return nil, errs.NotFound(errs.CodeNetworkNotFound, fmt.Sprintf("model.NetworkItem not found: %s", networkID))
	}
	if len(matches) > 1 {
		return nil, errs.NotFound(errs.CodeNetworkNotFound,
			fmt.Sprintf("model.NetworkItem ID is ambiguous: %s", networkID))
	}
	return r.enrich(ctx, matches)[0], nil
}

func (r *Resolver) ByName(ctx context.Context, name string, includeDeleted ...bool) (*model.NetworkItem, error) {
	network, err := r.repo.GetByName(ctx, name, includeDeleted...)
	if err != nil {
		return nil, err
	}
	if network == nil {
		return nil, errs.NotFound(errs.CodeNetworkNotFound, fmt.Sprintf("model.NetworkItem not found: %s", name))
	}
	return r.enrich(ctx, []*model.NetworkItem{network})[0], nil
}

func (r *Resolver) GetDefault(ctx context.Context) (*model.NetworkItem, error) {
	network, err := r.repo.GetDefault(ctx)
	if err != nil {
		return nil, err
	}
	if network == nil {
		return nil, nil
	}
	return r.enrich(ctx, []*model.NetworkItem{network})[0], nil
}

func (r *Resolver) Resolve(ctx context.Context, value string, includeDeleted ...bool) (*model.NetworkItem, error) {
	// Try by name first, then by ID prefix
	// from by_name — any other error (DB error, etc.) propagates immediately.
	network, err := r.ByName(ctx, value, includeDeleted...)
	if err == nil {
		return network, nil
	}
	if !errs.IsNotFound(err) {
		return nil, err // propagate non-not-found errors
	}
	network, err2 := r.ByID(ctx, value, includeDeleted...)
	if err2 == nil {
		return network, nil
	}
	// NOT the original by_name exception.
	return nil, err2
}

func (r *Resolver) ResolveMany(
	ctx context.Context,
	identifiers []string,
	includeDeleted ...bool,
) (*ResolveResult, error) {
	uniqueIDs := infra.Dedup(identifiers)

	var items []*model.NetworkItem
	var errorsList []string
	resolvedIDs := make(map[string]bool)

	for _, identifier := range uniqueIDs {
		item, err := r.Resolve(ctx, identifier, includeDeleted...)
		if err != nil {
			errorsList = append(errorsList, err.Error())
			continue
		}
		if !resolvedIDs[item.ID] {
			resolvedIDs[item.ID] = true
			items = append(items, item)
		}
	}

	items = r.enrich(ctx, items)

	exitCode := 0
	if len(errorsList) > 0 && len(items) == 0 {
		exitCode = 1
	}

	return &ResolveResult{
		Items:    items,
		Errors:   errorsList,
		ExitCode: exitCode,
	}, nil
}
